doc/coq.rst

```diff
@@ -388,6 +390,102 @@ documentation targets, we can use the ``@doc`` alias as in ``dune build @doc``.
 If we want to build all the LaTeX documentation then we use the ``@doc-latex``
 alias instead.
 
+.. _example plugin:
+
+Coq Plugin Project
+~~~~~~~~~~~~~~~~~~
+
+Let us build a simple Coq plugin to demonstrate how Dune can handle this setup. 
+
+.. code::
+
+  .
+  ├── dune-project
+  ├── src
+  │   ├── dune
+  │   ├── hello_world.ml
+  │   ├── my_plugin.mlpack
+  │   └── syntax.mlg
+  └── theories
+      ├── dune
+      └── UsingMyPlugin.v
+
+Our :ref:`dune-project` will need to have a package for the plugin to sit in,
+otherwise Coq will not be able to find it.
+
+.. code:: scheme
+
+  (lang dune 3.2)
+  (using coq 0.3)
+
+  (package
+   (name my-coq-plugin)
+   (synopsis "My Coq Plugin")
+   (depends coq-core))
+
+Now we have two directories, ``src/`` and ``theories/`` each with their own
+:ref:`dune file<dune-files>`. Let us begin with the plugin :ref:`dune
+file<dune-files>`:
+
+.. code:: scheme
+
+  (library
+   (name my_plugin)
+   (public_name my-coq-plugin.plugin)
+   (synopsis "My Coq Plugin")
+   (flags :standard -rectypes -w -27)
+   (libraries coq-core.vernac))
+
+  (coq.pp
+   (modules syntax))
+
+Here we define a library using the :ref:`library` stanza. Importantly, we
+declared which external libraries we rely on and gave the library a
+``public_name``, since this is how Coq will find it.
+
+The :ref:`coq-pp` stanza allows ``src/syntax.mlg`` to be preprocessed, which for
+reference looks like:
+
+.. code:: ocaml
+
+  DECLARE PLUGIN "my_plugin"
+
+  VERNAC COMMAND EXTEND CallToC CLASSIFIED AS QUERY
+  | [ "Hello" ] -> { Feedback.msg_notice Pp.(str Hello_world.hello_world) }
+  END
+
+Together with ``hello_world.ml``:
+
+.. code:: ocaml
+
+  let hello_world = "hello world!"
+
+They make up the plugin. There is one more important ingredient here, and that
+is the ``my_plugin.mlpack`` file. Only its existence is needed by ``coqdep``, so
+an empty file will suffice. See :ref:`this note on .mlpack
+files<limitation-mlpack>`.
+
+The file for ``theories/`` is a standard :ref:`coq-theory` stanza, with an
+included ``libraries`` field allowing Dune to see ``my-coq-plugin.plugin`` as a
+dependency.
+
+.. code:: scheme
+
+  (coq.theory
+   (name MyPlugin)
+   (package my-coq-plugin)
+   (libraries my-coq-plugin.plugin))
+
+Finally, our .v file will look something like this:
+
+.. code:: coq
+
+  Declare ML Module "my-coq-plugin.plugin".
```

Comment:
This has to be `my_plugin` in Coq < 8.16 or `my_plugin:my-coq-plugin.plugin` in Coq >= 8.16

When we implement runtime library deps we can drop the first part in Dune.